### PR TITLE
adding circuit breakers to envoy

### PR DESCRIPTION
**Pull request by GregHanson**
Addresses:
https://github.com/amalgam8/amalgam8/issues/476

### CODE REVIEW

File: sidecar/proxy/envoy/manager.go [hunk 1]
```diff
@@ -312,6 +313,20 @@ func buildClusters(rules []api.Rule) []Cluster {
 			ConnectTimeoutMs: 1000,
 		}
 
+		if backend.Timeout > 0 {
+			// convert from float to int
+			cluster.ConnectTimeoutMs = int(backend.Timeout * 1000)
+		}
```
Comment (rshriram) [inline sidecar/proxy/envoy/manager.go]: To avoid this format conversion, I suggest we change the rules schema to use timeout_budget_ms. What do you think?

**Comment by GregHanson** [comment]
@rshriram `timeout` and `retries` are in the rules schema under the `Backend`s field.  Should I add the remaining circuit breaker fields:

```
	MaxConnections    int `json:"max_connections,omitempty"`
	MaxPendingRequest int `json:"max_pending_requests,omitempty"`
	MaxRequests       int `json:"max_requests,omitempty"`
```

under the Backend struct as well?  Or should I make a substruct in Backend?  Not sure if adding the substruct in the rule schema will break any retry/timeout logic in the nginx lua code

File: sidecar/proxy/envoy/manager.go [hunk 2]
```diff
@@ -312,6 +313,20 @@ func buildClusters(rules []api.Rule) []Cluster {
...
+
+		if backend.Retries > 0 {
+			cluster.Hystrix = CircuitBreaker{
+				MaxRetries:        backend.Retries,
+				MaxConnections:    50,
+				MaxRequests:       25,
+				MaxPendingRequest: 5,
+			}
```
Comment (rshriram) [inline sidecar/proxy/envoy/manager.go]: hard coding?

**Review by rshriram — CHANGES_REQUESTED** [review]

File: sidecar/proxy/envoy/manager.go (hunk 2 above)
Comment (ijsnellf) [reply sidecar/proxy/envoy/manager.go]: See @GregHanson's comment above.

File: sidecar/proxy/envoy/manager.go (hunk 1 above)
Comment (ijsnellf) [reply sidecar/proxy/envoy/manager.go]: That will mean we have to do the inverse conversion in NGINX. Do we care about keeping NGINX up to date?

Comment (rshriram) [reply sidecar/proxy/envoy/manager.go]: no..

File: sidecar/proxy/envoy/manager.go [hunk 3]
```diff
@@ -292,24 +292,62 @@ func ParseServiceKey(s string) (string, []string) {
...
+			Hystrix:          CircuitBreaker{},
+			OutlierDetection: OutlierDetection{
+				MaxEjectionPercent: 100,
+			},
+		}
+
+		if backend.Hystrix != nil {
+			if backend.Hystrix.TimeoutMS > 0 {
+				// convert from float sec to int ms
+				cluster.ConnectTimeoutMs = int(backend.Hystrix.TimeoutMS * 1000)
```
Comment (ijsnellf) [inline sidecar/proxy/envoy/manager.go]: It is confusing that the field is called `timeout_ms`, but apparently contains float seconds. Either we should switch to a field name of just `timeout`, or we should have `timeout_ms` contain integer milliseconds.

File: pkg/api/rules-schema.json [hunk 4]
```diff
@@ -165,17 +165,65 @@
...
-        "retries": {
-          "description": "Number of times to retry a failed request to a different instances. Set to 0 to disable retries",
-          "type": "number",
-          "minimum": 0,
-          "exclusiveMinimum": false
+        "hystrix": {
+          "type": "object",
+          "properties": {
+            "timeout_ms": {
+              "description": "Connect/read/write timeout when talking to an instance. Default is nginx's timeout (60s). Unit is in seconds (e.g, 1.234s)",
```
Comment (ijsnellf) [inline pkg/api/rules-schema.json]: Remove reference to NGINX.

File: pkg/api/rules-schema.json [hunk 5]
```diff
@@ -165,17 +165,65 @@
...
+              "type": "number",
+              "minimum": 0,
+              "exclusiveMinimum": true
+            },
+            "max_retries": {
+              "description": "Maximum number of retries to connect to a backend",
```
Comment (ijsnellf) [inline pkg/api/rules-schema.json]: Mention defaults?

Comment (ijsnellf) [reply pkg/api/rules-schema.json]: Same comment for everything below...

File: sidecar/proxy/envoy/resources.go [hunk 6]
```diff
@@ -134,13 +135,29 @@ type Host struct {
...
-	MaxRequestsPerConnection int    `json:"max_requests_per_connection,omitempty"`
-	Hosts                    []Host `json:"hosts,omitempty"`
+	Name                     string           `json:"name"`
+	ServiceName              string           `json:"service_name,omitempty"`
+	ConnectTimeoutMs         int              `json:"connect_timeout_ms"`
+	Type                     string           `json:"type"`
+	LbType                   string           `json:"lb_type"`
+	MaxRequestsPerConnection int              `json:"max_requests_per_connection,omitempty"`
+	Hosts                    []Host           `json:"hosts,omitempty"`
+	Hystrix                  CircuitBreaker   `json:"circuit_breaker"`
```
Comment (ijsnellf) [inline sidecar/proxy/envoy/resources.go]: Field name should match the JSON tag to reduce confusion.

File: sidecar/proxy/envoy/resources.go [hunk 7]
```diff
@@ -70,6 +70,7 @@ type Route struct {
 	PrefixRewrite string   `json:"prefix_rewrite,omitempty"`
 	Cluster       string   `json:"cluster"`
 	Headers       []Header `json:"headers,omitempty"`
+	Timeout       string   `json:"timeout_ms,omitempty"`
```
Comment (ijsnellf) [inline sidecar/proxy/envoy/resources.go]: Was this added accidentally? I don't see it used anywhere, and it is type `string`.

File: sidecar/proxy/envoy/manager.go [hunk 8]
```diff
@@ -292,24 +292,62 @@ func ParseServiceKey(s string) (string, []string) {
...
 			Type:             "sds",
 			LbType:           "round_robin",
 			ConnectTimeoutMs: 1000,
+			Hystrix:          CircuitBreaker{},
+			OutlierDetection: OutlierDetection{
+				MaxEjectionPercent: 100,
+			},
+		}
+
+		if backend.Hystrix != nil {
```
Comment (ijsnellf) [inline sidecar/proxy/envoy/manager.go]: Put this in its own function?

File: sidecar/proxy/envoy/manager.go [hunk 9]
```diff
@@ -292,24 +292,62 @@ func ParseServiceKey(s string) (string, []string) {
 }
 
 func buildClusters(rules []api.Rule) []Cluster {
-	clusterMap := make(map[string]struct{})
+	clusterMap := make(map[string]*api.Backend)
```
Comment (ijsnellf) [inline sidecar/proxy/envoy/manager.go:295]: What about clusters with more than one backend?

**Review by ijsnellf — CHANGES_REQUESTED** [review]

File: sidecar/proxy/envoy/manager.go (hunk 3 above)
Comment (rshriram) [reply sidecar/proxy/envoy/manager.go]: I agree. We should change the json schema spec to timeout_ms and use timeout ms everywhere.

File: pkg/api/rules-schema.json (hunk 4 above)
Comment (rshriram) [reply pkg/api/rules-schema.json]: and say its in milliseconds

File: sidecar/proxy/envoy/manager.go (hunk 3 above)
Comment (GregHanson) [reply sidecar/proxy/envoy/manager.go]: @rshriram changing this mean changing all of our docs for bookinfo where we tell everyone to create a 7 second delay - still want me to go forward with that?

Comment (rshriram) [reply sidecar/proxy/envoy/manager.go]: On second thought, I take it back. Lets revert to using consistent "timeout".. and eliminate all the _ms and MS stuff. In an future PR, we can revisit changing all these to milliseconds.

**Comment by GregHanson** [comment]
Here is what a sample rule will look like with the new schema:

```
    {
      "id": "1e161ec1-b2ce-4518-a031-29ee56645918",
      "priority": 1,
      "destination": "productpage",
      "route": {
        "backends": [
          {
            "tags": [
              "version=v1"
            ],
            "resilience": {
              "max_connections": 4,
              "max_requests": 88,
              "consecutive_errors": 9,
              "max_requests_per_connection": 1021
            }
          }
        ],
        "tcp_connect_timeout": 0.5,
        "max_retries": 14
      }
    }
```

**Comment by GregHanson** [comment]
Here is what the sample envoy.json file will look like:

```
{
  "runtime": {
    "symlink_root": "/etc/envoy/runtime/routing",
    "subdirectory": "traffic_shift"
  },
  "listeners": [
    {
      "port": 6379,
      "filters": [
        {
          "type": "read",
          "name": "http_connection_manager",
          "config": {
            "codec_type": "auto",
            "stat_prefix": "ingress_http",
            "generate_request_id": true,
            "add_user_agent": true,
            "route_config": {
              "virtual_hosts": [
                {
                  "name": "backend",
                  "domains": [
                    "*"
                  ],
                  "routes": [
                    {
                      "runtime": {
                        "key": "reviews._:version=v1",
                        "default": 0
                      },
                      "prefix": "/reviews/",
                      "prefix_rewrite": "/",
                      "cluster": "reviews:version=v1",
                      "retry_policy": {
                        "retry_on": "5xx,connect-failure,refused-stream"
                      }
                    },
                    {
                      "runtime": {
                        "key": "productpage._:version=v1",
                        "default": 0
                      },
                      "prefix": "/productpage/",
                      "prefix_rewrite": "/",
                      "cluster": "productpage:version=v1",
                      "timeout_ms": 500,
                      "retry_policy": {
                        "retry_on": "5xx,connect-failure,refused-stream",
                        "num_retries": 14
                      }
                    }
              ]
            },
            "filters": [
              {
                "type": "decoder",
                "name": "router",
                "config": {
                  "dynamic_stats": false
                }
              }
            ],
            "access_log": [
              {
                "path": "/var/log/a8_access.log",
                "format": "{\"status\":\"%RESPONSE_CODE%\", \"start_time\":\"%START_TIME%\", \"request_time\":%DURATION%, \"upstream_response_time\":\"%RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%\", \"src\":\"gateway:version=none\", \"dst\":\"%UPSTREAM_CLUSTER%\", \"gremlin_recipe_id\":\"-\", \"message\":\"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %RESPONSE_CODE%\", \"module\":\"ENVOY\", \"method\":\"%REQ(:METHOD)%\", \"path\":\"%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%\", \"protocol\":\"%PROTOCOL%\", \"response_flags\":\"%RESPONSE_FLAGS%\", \"x_forwarded\":\"%REQ(X-FORWARDED-FOR)%\", \"user_agent\":\"%REQ(USER-AGENT)%\", \"request_id\":\"%REQ(X-REQUEST-ID)%\", \"auth\":\"%REQ(:AUTHORITY)%\", \"upstream_host\":\"%UPSTREAM_HOST%\"}\n"
              }
            ]
          }
        }
      ]
    }
  ],
  "admin": {
    "access_log_path": "/var/log/envoy_admin.log",
    "port": 8001
  },
  "cluster_manager": {
    "clusters": [
      {
        "name": "details:version=v1",
        "service_name": "details:version=v1",
        "connect_timeout_ms": 1000,
        "type": "sds",
        "lb_type": "round_robin",
        "circuit_breaker": {},
        "outlier_detection": {
          "max_ejection_percent": 100
        }
      },
      {
        "name": "productpage:version=v1",
        "service_name": "productpage:version=v1",
        "connect_timeout_ms": 1000,
        "type": "sds",
        "lb_type": "round_robin",
        "max_requests_per_connection": 1021,
        "circuit_breaker": {
          "max_connections": 4,
          "max_requests": 88
        },
        "outlier_detection": {
          "consecutive_5xx": 9,
          "max_ejection_percent": 100
        }
      }
    ],
    "sds": {
      "cluster": {
        "name": "sds",
        "connect_timeout_ms": 1000,
        "type": "strict_dns",
        "lb_type": "round_robin",
        "max_requests_per_connection": 1,
        "hosts": [
          {
            "url": "tcp://127.0.0.1:6500"
          }
        ]
      },
      "refresh_delay_ms": 1000
    }
  }
}
```

Note that productpage `route` and `cluster` have the fields configured while the other services demonstrate the default settings

File: pkg/api/rule.go [hunk 10]
```diff
@@ -41,7 +41,9 @@ type Match struct {
 
 // Route definition
 type Route struct {
-	Backends []Backend `json:"backends"`
+	Backends          []Backend `json:"backends"`
+	TCPConnectTimeout float64   `json:"tcp_connect_timeout,omitempty"`
```
Comment (rshriram) [inline pkg/api/rule.go]: HttpReqTimeout

File: pkg/api/rule.go [hunk 11]
```diff
@@ -41,7 +41,9 @@ type Match struct {
...
+	MaxRetries        int       `json:"max_retries,omitempty"`
 }
```
Comment (rshriram) [inline pkg/api/rule.go]: HttpReqRetries

File: pkg/api/rules-schema.json [hunk 12]
```diff
@@ -192,6 +228,18 @@
           "items": {
             "$ref": "#/definitions/backend"
           }
+        },
+        "tcp_connect_timeout": {
+          "description": "Connect/read/write timeout when talking to a route. Default is 15. Unit is in seconds.",
```
Comment (rshriram) [inline pkg/api/rules-schema.json]: http_req_timeout
Timeout for a HTTP request. Includes retries as well. Unit is in floating point seconds. Default 15.0s

File: pkg/api/rules-schema.json [hunk 13]
```diff
@@ -192,6 +228,18 @@
...
+          "type": "number",
+          "minimum": 0,
+          "exclusiveMinimum": true
+        },
+        "max_retries": {
+          "description": "Maximum number of retries to connect to a route.  Defaults to 1.",
```
Comment (rshriram) [inline pkg/api/rules-schema.json]: http_req_retries
Number of retries for a given request.  The interval between retries will be determined automatically (25ms+). Actual number of retries attempted depends on the http_req_timeout

**Review by rshriram — CHANGES_REQUESTED** [review]

**Review by rshriram — APPROVED** [review]

File: sidecar/proxy/envoy/manager.go [hunk 14]
```diff
@@ -292,24 +292,60 @@ func ParseServiceKey(s string) (string, []string) {
...
 			Type:             "sds",
 			LbType:           "round_robin",
 			ConnectTimeoutMs: 1000,
+			CircuitBreaker:   &CircuitBreaker{},
+			OutlierDetection: &OutlierDetection{
+				MaxEjectionPercent: 100,
+			},
+		}
+
+		if backend.Resilience != nil {
```
Comment (ijsnellf) [inline sidecar/proxy/envoy/manager.go]: No function? That's ok. That's fine.

**Pull request merged**